authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,147 @@
 #  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  * See the License for the specific language governing permissions and
 #  * limitations under the License.
-
+import logging
 
 import ldap
 from flask_securest.authentication_providers.abstract_authentication_provider \
     import AbstractAuthenticationProvider
 from flask_securest import utils
+from flask_securest import constants
+from flask_securest.exceptions import AuthenticationException
+
+DN_ATTRIBUTE_NAME = 'DN'
+DEFAULT_AUTH_ATTR_NAME = 'uid'
 
 
 class LDAPAuthenticationProvider(AbstractAuthenticationProvider):
-    def __init__(self, directory_url):
-        self.directory_url = directory_url
+    def __init__(self,
+                 ldap_url,
+                 base_dn=None,
+                 admin_user_id=None,
+                 admin_password=None,
+                 user_id_attribute=None):
+        self.lgr = logging.getLogger(constants.FLASK_SECUREST_LOGGER_NAME)
+        self.ldap_url = ldap_url
+        self.base_dn = base_dn
+        self.admin_user_id = admin_user_id
+        self.admin_password = admin_password
+        self.user_id_attribute = user_id_attribute
 
+    # LDAP authentication can be done by subtree match or according to
+    # dn attribute.
     def authenticate(self, userstore=None):
-        username, password = utils.get_basic_http_authentication_info()
+        who, password = self.get_creds_from_request()
+        if self.is_subtree_authentication_method():
```

Comment:
1. the authentication method is not sub-tree, it's by user-binding, so it should be renamed
2. it should work even if no sub-tree was supplied, I know that's not the example Uri gave you but generally if no sub-tree is given the root is expected to be used, that's the default